### DISCUSSION TITLE
Using cocotb without HDL for other standalone event based simulations

### DISCUSSION THREAD

**Discussion by eehusky** [cocotb/cocotb#3889] [Q&A] [answered]
After having used cocotb for awhile (the way its intended to be used) I've found myself wanting to use the same sort of asynchronous semantics for some other event based simulation work that has absolutely nothing to do with HDL/FPGAs/RTL and their ilk.  

Is there anyway to setup a cocotb environment like this?  Short of creating a dummy dut that doesn't do anything and running a simulation against that, I'm fine with going that route but it would be nice to be able to remove the dependency and all the supporting cruft that goes along with it (and maybe get a performance boost?).

**Comment by ktbarrett** [comment]
You might be looking for [`asyncio`](https://docs.python.org/3/library/asyncio.html) or [`trio`](https://trio.readthedocs.io/en/stable/). If you also need a testing framework, I suggest [`pytest`](https://docs.pytest.org/en/8.2.x/). There are pytest plugins that allow you to use asyncio, such as [`pytest-asyncio`](https://pytest-asyncio.readthedocs.io/en/latest/) or [`pytest-trio`](https://pytest-trio.readthedocs.io/en/stable/).

**Reply by eehusky** [reply]
ah yup, well thats sort of why I ended up asking the question.  

I found myself recreating the same sort scheduling/triggering stuff that is all baked into cocotb already (along with the existing test discovery integration) and was thinking, well, if i could just remove/ignore the hdl specific stuff from cocotb i wouldn't have to redo all this.  

Ive never had a reason to dig into the gpi side of a simulator but i was thinking about making something along the lines making a stub simulator that implements whats in simluator.pyi that handles just advancing time and signal read/write callbacks.  But that sounds like it could be dangerously close to just as much futzing around as recreating the bits of cocotb i'd like to make use of (and now that i say it out loud.....not so much different than using a dummy verilated model.)

**Reply by ktbarrett** [reply, marked as answer]
I guess I don't understand then. `pytest` and `asyncio` do everything cocotb does (and more) if you rip out everything related to the simulator.

Do you mean you want a concurrency framework with discrete time steps?

**Reply by eehusky** [reply]
Yeah!  That sounds exactly like what I'm after!  Sort of like what SimPy does...but not (the last time i tried it out it was kinda clunky, might have to take another look with a fresh set of eyes).

May not be the best analogy, but, sans the graphical part, the combination of the cocotb concurrency/scheduling bits and everything else you can attach to it via python is the intuitive-ist replacement I've found to using simulink.  But its sort of backwards in the sense of simulink being inherently general purpose but lets you connect to and interact with hdl simulations.  

Like if you had a "simulator" attached to cocotb that was just a scheduling kernel with a sort of signal database attached to it.

**Reply by ktbarrett** [reply]
I think you could make an object that works with `asyncio` or `trio` that adds discrete stepping.

**Reply by cmarqu** [reply]
There is also https://usim.readthedocs.io/en/latest/index.html but the project appears dormant.

**Reply by cmarqu** [reply]
Another such project: https://crestdsl.readthedocs.io/

**Reply by themperek** [reply]
Maybe https://github.com/fvutils/cocotb-stub-sim (https://github.com/themperek/cocotb-vivado is based on this)

**Reply by eehusky** [reply]
That looks exactly like what i was looking for!  Thanks!

**Reply by cmarqu** [reply]
NKOTB: https://docs.pathsim.org/
